src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Runtime.CompilerServices
+{
+    internal static partial class AsyncMethodBuilder
+    {
+        /// <summary>Initiates the builder's execution with the associated state machine.</summary>
+        /// <typeparam name="TStateMachine">Specifies the type of the state machine.</typeparam>
+        /// <param name="stateMachine">The state machine instance, passed by reference.</param>
+        [DebuggerStepThrough]
+        public static void Start<TStateMachine>(ref TStateMachine stateMachine) where TStateMachine : IAsyncStateMachine
```

Comment:
@stephentoub @benaadams This method will get stamped for every TStateMachine. I do not think I can bring this change over as is to ProjectN TFS. I expect that it would cause pretty measurable compiled binary size regression there. There was significant work done in the CoreRT AsyncMethodBuilder to minimize size of TStateMachine stamps, e.g. https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs#L736 . It is one of the reasons why the AsyncMethodBuilder is quite different in CoreRT.

Could you please make this change under ifdef, or roll it back ?

This probably change likely hurts CoreCLR startup/footprint as well. Ideally, we would get the AsyncMethodBuilder in sync so that we get the footprint optimizations in CoreCLR; and also make this throughput optimization compatible with the footprint optimizations.

Reply:
It can be changed, but is it really going to be that much of a difference?  Today this logic just lives in `AsyncTaskMethodBuilder<T>.Start`, and now `AsyncTaskMethodBuilder<T>.Start` is a one-liner that calls to the logic moved to `AsyncMethodBuilder.Start<T>`.  Is the concern the extra one-liner wrapper?  Or a concern about the logic being in a generic method rather than in a method on a generic type?  Or I guess it's that because CoreRT already has its own copy that this doesn't affect, so now there's two of them?  As it stands this logic will only be compiled in if any methods in CoreRT have "async ValueTask<T>" methods.

Reply:
I'd be very happy for coreclr to drop the finally block as per corert (raised issue for it https://github.com/dotnet/coreclr/issues/11156); but I don't think a decision was taken /cc @stephentoub

Reply:
> Ideally, we would get the AsyncMethodBuilder in sync so that we get the footprint optimizations in CoreCLR

FYI, the async builder in coreclr has basically been completely rewritten (previously).

Reply:
`ExecutionContextSwitcher` gets inlined on the corert version, so the main disparity is from the try/finally

Reply:
https://github.com/dotnet/coreclr/pull/16124

Reply:
> so the main disparity is from the try/finally

I don't think that's the sole difference Jan is referring to.

Reply:
> There was significant work done in the CoreRT AsyncMethodBuilder to minimize size of TStateMachine stamps, e.g. `internal static Action GetCompletionAction<TStateMachine>`

This is changing the method just above `Start<TStateMachine>(ref TStateMachine stateMachine)`
https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs#L724

Which as I understand it has [ExecutionContext.EstablishCopyOnWriteScope](https://github.com/dotnet/corert/blob/cc2ed91493c6191d316c9805ee2bc2c2d1369afe/src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs#L173-L179) inlined and [ExecutionContextSwitcher.Undo](https://github.com/dotnet/corert/blob/cc2ed91493c6191d316c9805ee2bc2c2d1369afe/src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs#L29-L43) inlined - which has the same flow but lives via stack rather than registers

Reply:
Added extra changes for corert https://github.com/dotnet/corert/pull/5324

Reply:
Reduced code size in coreclr https://github.com/dotnet/coreclr/pull/16129